astropy/utils/masked/core.py

```diff
@@ -0,0 +1,1086 @@
+        ----------
+        data : array-like
+            Possibly masked item, judged by whether it has a ``mask`` attribute.
+            If so, checks for being an instance of `~astropy.utils.masked.Masked`
+            or `~numpy.ma.MaskedArray`, and gets unmasked data appropriately.
+
+        Returns
+        -------
+        unmasked, mask : array-like
+            Unmasked will be `None` for `~numpy.ma.masked`.
+
+        """
+        mask = getattr(data, 'mask', None)
+        if mask is not None:
+            try:
+                data = data.unmasked
+            except AttributeError:
+                if not isinstance(data, np.ma.MaskedArray):
+                    raise
+                if data is np.ma.masked:
+                    if allow_ma_masked:
+                        data = None
+                    else:
+                        raise ValueError('cannot handle np.ma.masked here.') from None
+                else:
+                    data = data.data
+
+        return data, mask
+
+    @classmethod
+    def _data_masks(cls, *args):
+        data_masks = [cls._data_mask(arg) for arg in args]
+        return (tuple(data for data, _ in data_masks),
+                tuple(mask for _, mask in data_masks))
+
+    def _get_mask(self):
+        """The mask.
+
+        If set, replace the original mask, with whatever it is set with,
+        using a view if no broadcasting or type conversion is required.
+        """
+        return self._mask
+
+    def _set_mask(self, mask, copy=False):
+        self_dtype = getattr(self, 'dtype', None)
+        mask_dtype = (np.ma.make_mask_descr(self_dtype)
+                      if self_dtype and self_dtype.names else np.dtype('?'))
+        ma = np.asanyarray(mask, dtype=mask_dtype)
+        if ma.shape != self.shape:
+            # This will fail (correctly) if not broadcastable.
+            self._mask = np.empty(self.shape, dtype=mask_dtype)
+            self._mask[...] = ma
+        elif ma is mask:
+            # Even if not copying use a view so that shape setting
+            # does not propagate.
+            self._mask = mask.copy() if copy else mask.view()
+        else:
+            self._mask = ma
+
+    mask = property(_get_mask, _set_mask)
+
+    # Note: subclass should generally override the unmasked property.
+    # This one assumes the unmasked data is stored in a private attribute.
+    @property
+    def unmasked(self):
+        """The unmasked values."""
+        return self._unmasked
+
+    def filled(self, fill_value=None):
+        """Get the underlying data, possibly filling masked values."""
+        unmasked = self.unmasked
+        if fill_value is None:
+            return unmasked
+        else:
+            unmasked = unmasked.copy()
+            if self.mask.dtype.names:
+                np.ma.core._recursive_filled(unmasked, self.mask, fill_value)
+            else:
+                unmasked[self.mask] = fill_value
+            return unmasked
```

Comment:
All of 2 lines shorter, but it just makes more evident what is the test case

```suggestion
        if fill_value is not None:
            unmasked = unmasked.copy()
            if self.mask.dtype.names:
                np.ma.core._recursive_filled(unmasked, self.mask, fill_value)
            else:
                unmasked[self.mask] = fill_value
        return unmasked
```

Reply:
Actually, looking at this again, I think really the argument should just be obligatory - if only so that one always gets a copy. But also because unlike `np.ma.MaskedArray`, this class does not carry a fill value around so there is no default. And this has the advantage that one can also fill an object array with `None`.

Reply:
p.s. This is a leftover from when `unmasked = property(filled)`...

Reply:
Sounds good. What's the way to close a suggestion? They always show up for me even when I resolve them.

Reply:
Weird, for me they disappear. Does one have to be author or member perhaps?